core/src/main/scala/org/apache/spark/deploy/rest/StandaloneRestClient.scala

```diff
@@ -63,45 +71,109 @@ private[deploy] class StandaloneRestClient extends Logging {
-        reportSubmissionStatus(master, s)
-        handleRestResponse(s)
-      case unexpected =>
-        handleUnexpectedRestResponse(unexpected)
+    var suc: Boolean = false
+    var response: SubmitRestProtocolResponse = null
+    for (m <- masters if !suc) {
+      validateMaster(m)
+      val url = getSubmitUrl(m)
+      try {
+        response = postJson(url, request.toJson)
+        response match {
+          case s: CreateSubmissionResponse =>
+            if (s.success) {
+              reportSubmissionStatus(s)
+              handleRestResponse(s)
+              suc = true
+            }
+          case unexpected =>
+            handleUnexpectedRestResponse(unexpected)
+        }
+      } catch {
+        case unreachable @ (_: FileNotFoundException | _: SocketException | _: ConnectException) =>
+          if(handleConnectionException(m)) {
+            throw new SubmitRestConnectionException(
+              s"Unable to connect to server", unreachable)
+          }
+        case malformed @ (_: JsonProcessingException | _: SubmitRestProtocolException) =>
+          if(handleConnectionException(m)) {
+            throw new SubmitRestProtocolException(
+              "Malformed response received from server", malformed)
+          }
+      }
     }
     response
   }
 
   /** Request that the server kill the specified submission. */
-  def killSubmission(master: String, submissionId: String): SubmitRestProtocolResponse = {
+  def killSubmission(submissionId: String): SubmitRestProtocolResponse = {
     logInfo(s"Submitting a request to kill submission $submissionId in $master.")
-    validateMaster(master)
-    val response = post(getKillUrl(master, submissionId))
-    response match {
-      case k: KillSubmissionResponse => handleRestResponse(k)
-      case unexpected => handleUnexpectedRestResponse(unexpected)
+    var suc: Boolean = false
+    var response: SubmitRestProtocolResponse = null
+    for (m <- masters if !suc) {
+      validateMaster(m)
+      val url = getKillUrl(m, submissionId)
+      try {
+        response = post(url)
+        response match {
+          case k: KillSubmissionResponse =>
+            if (!k.message.startsWith(Utils.MASTER_NOT_ALIVE_STRING)) {
+              handleRestResponse(k)
+              suc = true
+            }
+          case unexpected =>
+            handleUnexpectedRestResponse(unexpected)
+        }
+      } catch {
+        case unreachable @ (_: FileNotFoundException | _: SocketException | _: ConnectException) =>
+          if(handleConnectionException(m)) {
+            throw new SubmitRestConnectionException(
+              s"Unable to connect to server", unreachable)
+          }
+        case malformed @ (_: JsonProcessingException | _: SubmitRestProtocolException) =>
+          if(handleConnectionException(m)) {
+            throw new SubmitRestProtocolException(
+              "Malformed response received from server", malformed)
+          }
+      }
```

Comment:
If the concern is that you need to pass the master URL to `handleConnectionException`, you can instead pass the actual URL (i.e. without the spark:// prefix, which `readResponse` does have).